### PR TITLE
[mqtt.generic] Support chaining transformations without an intersection symbol

### PR DESCRIPTION
**Pull request by jimtng**
- Refactored by removing the transformation handling code from this binding and using core's `ChannelTransformation` class.

- Because core's class [supports transformation syntax with parentheses](https://github.com/openhab/openhab-core/pull/4352) `JSONPATH($.path)` as well as colon `JSONPATH:$.path`, it automatically extends that feature to this binding.

- Supports chaining multiple transformations by setting the channel configuration to `multiple="true"`. This results in the [transformations being held in a list](https://github.com/openhab/openhab-core/pull/4353). The old intersection symbol is also supported and gets combined into the list.

The build currently fails because this depends on https://github.com/openhab/openhab-core/pull/4355

### CODE REVIEW

**Comment by lsiepel** [comment]
As the previous PR is closed, my question might get lost:
The CHAIN transformation profile is about to get merged, that looks more like a generic solution for all bindings. Is there a reason that mqtt (and http) need to keep there own chaining mechanism?

For reference: https://github.com/openhab/openhab-addons/pull/17192

Edit: Not directly related to this PR, but it triggered me to comment.

**Comment by jimtng** [comment]
> The CHAIN transformation profile is about to get merged, that looks more like a generic solution for all bindings. Is there a reason that mqtt (and http) need to keep there own chaining mechanism?

- backwards compatibility. 
- doing the transformation on the binding once for many consumers (items), vs the profile which must be applied to every link
- This PR + http + more to come is about removing all custom chaining code and replacing them with a single implementation provided by core. The bindings merely provide the UI via their configurations.

**Comment by J-N-K** [comment]
Is this still a draft?

**Comment by jimtng** [comment]
It's a draft because it's waiting for the core pr linked above

**Comment by J-N-K** [comment]
Sorry, completely forgot about that.

**Comment by jimtng** [comment]
rebased

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelConfig.java [hunk 1]
```diff
@@ -47,8 +48,8 @@ public class ChannelConfig {
     public boolean trigger = false;
     public String unit = "";
 
-    public String transformationPattern = "";
-    public String transformationPatternOut = "";
+    public @Nullable List<String> transformationPattern;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelConfig.java:51]: Is there a reason to have this `@Nullable`? Wouldn't `public List<String> transformationPattern = List.of()` also work?

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/handler/GenericThingConfiguration.java [hunk 2]
```diff
@@ -42,5 +44,5 @@ public class GenericThingConfiguration {
     /**
      * transformation pattern for the availability payload
      */
-    public @Nullable String transformationPattern;
+    public @Nullable List<String> transformationPattern;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/handler/GenericThingConfiguration.java:47]: see above

**Review by J-N-K — CHANGES_REQUESTED** [review]
Many thanks for the cleanup. In general LGTM. Just some minor comments.

File: bundles/org.openhab.binding.mqtt.generic/README.md [hunk 3]
```diff
@@ -285,20 +308,18 @@ This is required if your received value is wrapped in a JSON or XML response.
...
-
-Transformations can be chained by separating them with the mathematical intersection character "∩".
-Please note that the incoming value will be discarded if one transformation fails (e.g. REGEX did not match).
+| Received value                                                      | Tr. Service | Transformation                             |
+|---------------------------------------------------------------------|-------------|--------------------------------------------|
+| `{device: {status: { temperature: 23.2 }}}`                         | JSONPATH    | `JSONPATH($.device.status.temperature)`    |
+| `<device><status><temperature>23.2</temperature></status></device>` | XPath       | `XPath(/device/status/temperature/text())` |
+| `THEVALUE:23.2°C`                                                   | REGEX       | `REGEX(:(.*?)°)`                           |
+| `abc`                                                               | JS          | `JS(to_uppercase.js)`                      |
+| `abc`                                                               | JS (inline) | `JS(\| input.toUpperCase() )`              |
```
[the remark this answers is not in the document]
Comment (jimtng) [reply bundles/org.openhab.binding.mqtt.generic/README.md]: I've added a few more examples

File: bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelConfig.java (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelConfig.java]: I've changed it to non nullable

**Review by J-N-K — APPROVED** [review]
LGTM.

**Comment by J-N-K** [comment]
@ccutrer @antroids Any comments? Otherwise I'll merge tomorrow morning (CET).

**Review by ccutrer — APPROVED** [review]

**Pull request merged**